File: benchmarks/tsb/bench_cat_remove_unused.ts
/**
 * Benchmark: cat_remove_unused — CategoricalAccessor.removeUnusedCategories() on 100k-element Series
 */
import { Series } from "../../src/index.js";

const ROWS = 100_000;
const WARMUP = 3;
const ITERATIONS = 10;

const cats = ["a", "b", "c"];
const base = new Series({ data: Array.from({ length: ROWS }, (_, i) => cats[i % cats.length]) });
// Add extra categories that are unused so removeUnusedCategories has work to do
const s = base.cat.addCategories(["x", "y", "z"]);

for (let i = 0; i < WARMUP; i++) {
  s.cat.removeUnusedCategories();
}

const start = performance.now();
for (let i = 0; i < ITERATIONS; i++) {
  s.cat.removeUnusedCategories();
}
const total = performance.now() - start;

console.log(
  JSON.stringify({
    function: "cat_remove_unused",
    mean_ms: total / ITERATIONS,
    iterations: ITERATIONS,
    total_ms: total,
  }),
);

File: benchmarks/pandas/bench_cat_remove_unused.py
"""Benchmark: cat_remove_unused — pd.Categorical.remove_unused_categories() on 100k-element Series"""
import json, time
import numpy as np
import pandas as pd

ROWS = 100_000
WARMUP = 3
ITERATIONS = 10

cats = ["a", "b", "c"]
data = [cats[i % len(cats)] for i in range(ROWS)]
# Add unused categories
cat_type = pd.CategoricalDtype(categories=["a", "b", "c", "x", "y", "z"])
s = pd.Series(data, dtype=cat_type)

for _ in range(WARMUP):
    s.cat.remove_unused_categories()

start = time.perf_counter()
for _ in range(ITERATIONS):
    s.cat.remove_unused_categories()
total = (time.perf_counter() - start) * 1000

print(json.dumps({
    "function": "cat_remove_unused",
    "mean_ms": total / ITERATIONS,
    "iterations": ITERATIONS,
    "total_ms": total,
}))
